Implement memory range mapping/unmapping; Cleanup and standardize vmap API

wipawel opened the pull request:
This resolves #253.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+static int _vunmap(cr3_t *cr3_ptr, void *va) {
+    pgentry_t *tab;
+
+    if (!va)
```
Unmapping, and as a result, mapping address zero, i.e. `NULL` isn't supported by design?

wipawel replied:
Yes, as of now we assume `0`/`NULL` to be invalid virtual address.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+    if (!va)
+        return false;
+
+    dprintk("%s: va: %p (cr3: 0x%p)\n", __func__, va, cr3_ptr);
```
The "0x" prefix in the format string is redundant for printing `cr3_ptr`?

wipawel replied:
The `%p` does not add `0x` prefix in the implementation we use. I will add the `0x` to the `va` part too.

sktt replied:
Alternatively, use`%016x`, in case `%p` one day adds the "0x" prefix?

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+
+#define MMAP_RANGE_SIZE(size, mfn, flags, mmap_flags)                                    \
+    do {                                                                                 \
+        if (likely((mmap_flags) &MMAP_KERNEL))                                           \
```
Please add a space between the operator and the constant to ease readability, i.e. `… & MMAP_KERNEL`.
Same for all the others below.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -33,6 +34,9 @@
 cr3_t user_cr3;
 
+static spinlock_t vmap_lock = SPINLOCK_INIT;
+static spinlock_t mmap_lock = SPINLOCK_INIT;
```
Might be worth to document the locking order: `mmap_lock` before `vmap_lock` and not vice versa to avoid deadlocks via concurrent users.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+    spin_lock(&vmap_lock);
+
+    if (mfn_invalid(cr3_ptr->mfn)) {
+        spin_unlock(&vmap_lock);
+        return -EINVAL;
```
What about having that `spin_unlock(&vmap_lock); return -E...;` pattern to be transformed into `err = -E...; goto unlock_out;`? This would lower the amount of exit paths (to review) to just a few, making it easier to reason about the locking state.

wipawel replied:
Ok

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+
+    set_pgentry(&l1e->entry, MFN_INVALID, PT_NO_FLAGS);
+    spin_unlock(&vmap_lock);
+    return 0;
```
This would change to:
```c
unlock_out:
    spin_unlock(&vmap_lock);
    return err;
```

With `err` being a local variable initialized to 0.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+            vmap_##size(mfn_to_virt_user((mfn)), (mfn), (flags & ~_PAGE_USER));          \
+        if ((mmap_flags) &MMAP_KERNEL_MAP)                                               \
+            vmap_##size(mfn_to_virt_map((mfn)), (mfn), (flags));                         \
+    } while (0)
```
Should this be an `if (...) else if (...) else if (...)` ladder or even a switch block instead to avoid mapping regions multiple times in case `flags` is a combination of multiple, e.g. `MMAP_KERNEL | MMAP_USER`?

Also, the code is completely ignoring the return value of the kmap/vmap calls.

wipawel replied:
It is actually by design to support multiple mappings into different areas. Quite often we need to map the same region into VIRT_KERNEL_BASE area and/or VIRT_USER_BASE (or IDENT).

Good point about the return values. The only thing that makes sense here (since the return values are mainly ignored later one) is to `BUG_ON(NULL)` I think.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+    paddr_t end = cur + size;
+
+    /* Round up to the next page boundary unless it is page aligned */
+    end = (end + PAGE_SIZE - 1) & PAGE_MASK;
```
Check for overflows, maybe?
```c
if (end < start)
    return NULL;
```

wipawel replied:
Ok.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+void *mmap_range(paddr_t paddr, size_t size, unsigned long flags,
+                 mmap_flags_t mmap_flags) {
+    paddr_t cur = paddr;
+    paddr_t end = cur + size;
```
As the *map functions require to be passed a properly aligned address, should the code enforce this too? Or is this part of the contract, i.e. the callers responsibility?

Either way, we should check (and possibly handle) non-page-aligned `cur`.

wipawel replied:
I am not sure I understand fully. The *map functions take two parameters:
1. where to map (the `va`)
2. what to map (the `mfn`)
They have to check if the `va` is properly aligned for the mapping according to the order of the `mfn`.

Here, the `mmap_range()` takes a physical address and turns it into a `mfn`, so it will automatically take care of the non-page-aligned addresses. It maps it automatically to the aligned addresses using various `mfn_to_virt_*()` helpers.

minipli-oss replied:
Ahh, right. The conversion to a page frame number for the *map() calls implicitly aligns `cur` and the conversion at the end using `paddr_to_virt_map()` will preserve any possible offset.

Sorry for the noise, somehow overlooked the `paddr_to_mfn()` in the macros maze :D

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+done_2m:
+    for (; cur < end; cur += PAGE_SIZE)
+        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
+
```
Can't these multiple first ascending, than descending for-loops be simplified to a single loop like this?:

```c
while (size >= PAGE_SIZE) {
    if (size >= PAGE_SIZE_1G && !(cur % PAGE_SIZE_1G)) {
        MMAP_RANGE_1G(paddr_to_mfn(cur), flags, mmap_flags);
        cur += PAGE_SIZE_1G;
        size -= PAGE_SIZE_1G;
        continue;
    }

    if (size >= PAGE_SIZE_2M && !(cur % PAGE_SIZE_2M)) {
        MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
        cur += PAGE_SIZE_2M;
        size -= PAGE_SIZE_2M;
        continue;
    }

    MMAP_RANGE_4k(paddr_to_mfn(cur), flags, mmap_flags);
    cur += PAGE_SIZE;
    size -= PAGE_SIZE;
}
```

This should achieve the same thing (using the best fitting mapping for a given size/alignment) with less code.

wipawel replied:
Ok, good idea... the following seems to work fine:

```c
    while (cur < end) {
        if (size >= PAGE_SIZE_1G && !(cur % PAGE_SIZE_1G)) {
            MMAP_RANGE_1G(paddr_to_mfn(cur), flags, mmap_flags);
            cur += PAGE_SIZE_1G;
            size -= PAGE_SIZE_1G;
            continue;
        }

        if (size >= PAGE_SIZE_2M && !(cur % PAGE_SIZE_2M)) {
            MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
            cur += PAGE_SIZE_2M;
            size -= PAGE_SIZE_2M;
            continue;
        }

        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
        cur += PAGE_SIZE;
        size -= PAGE_SIZE;
    }
```

minipli-oss replied:
Should be equivalent to what I proposed, yes.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+    else if (mmap_flags & MMAP_USER)
+        return paddr_to_virt_user(paddr);
+    else if (mmap_flags & MMAP_IDENT)
+        return paddr_to_virt(paddr);
```
Ahh, here it's an if-else-if ladder, so should probably be one in `MMAP_RANGE_SIZE()` too.

This one is missing `MMAP_USER_KERNEL`, tho. So, maybe, using a `switch` block without a default label would be better, as it would warn about unhandled cases?:

```c
switch (mmap_flags) {
case MMAP_KERNEL_MAP:
    return paddr_to_virt_map(paddr);
case MMAP_KERNEL:
    return paddr_to_virt_kern(paddr);
case MMAP_USER:
    return paddr_to_virt_user(paddr);
case MMAP_IDENT:
    return paddr_to_virt(paddr);
}

return NULL;
```

However, the `mmap_flags` sanity check should probably be done at the begin of the function.

wipawel replied:
Good point about the missing `MMAP_USER_KERNEL` and about turning it into a `switch` block. Will do.

But, the functionality I am trying to implement here is to be able to map given `mfn` to whatever memory region selected via the `mmap_flags`. This takes care of the KERNEL, IDENT and USER regions for the kernel pagetable and for the USER and KERNEL (called USER_KERNEL) for the user pagetable.
So we want to be able to map all selected regions with a one call and swoop. 
However, what this function returns for multiple regions is somewhat pre-defined as an implicit contract. It is defined by this if-else ladder. Essentially, callers receive the "top-most" memory region `va` address and if they want to use other region `va` from the ones specified, they have to calculate it on their own.

minipli-oss replied:
If `mmap_flags` can be actually multiple flags, using a `switch` block won't work.

Returning the "highest" virtual address is also kinda fuzzy. I mean, the callers should know what flags they passed and thereby what address they should expect to get. But still, feels a little fragile to me... But if it orks for you, I shouldn't complain!

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+        if (mmap_flags & MMAP_USER_KERNEL)
+            vunmap_kern(paddr_to_virt_user(cur));
+        if (mmap_flags & MMAP_KERNEL_MAP)
+            vunmap_kern(paddr_to_virt_map(cur));
```
Same if-else-if / switch argumentation as for `MMAP_RANGE_SIZE()`

wipawel replied:
Same reasoning as above applies here: we want to be able to unmap multiple regions with one call.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
-                vmap_4k(va, mfn, r->flags);
-                vmap_user_4k(va, mfn, r->flags);
-            }
+            flags = MMAP_USER | MMAP_USER_KERNEL;
```
Oh, a combination of flags is actually a valid use case? If so, either the 'if (flag & FOO) ...; if (flag & BAR) ...;` pattern is broken or the if-else-if-else-if one is.

wipawel replied:
Yes it is. We want to be able to map the `mfn` to either multiple memory regions or kernel and user pagetable or a combination thereof. I think the confusing part is what the `mmap_range()` returns. Maybe the previous comment above clarifies it a bit?

minipli-oss commented on common/acpi.c:
```diff
@@ -134,29 +134,12 @@ static rsdp_rev1_t *acpi_find_rsdp(void) {
-
-    return num_pages;
+static void *acpi_table_map_pages(paddr_t pa, size_t len) {
+    return mmap_range(pa, len, L1_PROT, MMAP_KERNEL);
```
The old code handled non-page aligned `pa`, maybe `mmap_range()` should too?

wipawel replied:
I think it still does so.

minipli-oss replied:
You're right, I was misreading `mmap_range()`'s implementation.

minipli-oss commented on common/usermode.c:
```diff
@@ -237,6 +234,6 @@ void *__user_text mmap(void *va, unsigned long order) {
 
-void __user_text munmap(void *va, unsigned long order) {
-    sys_munmap(va, order);
+int __user_text munmap(void *va) {
+    return sys_munmap(va);
```
This change is unfortunate, as this will only unmap the very first page. :(

What about passing a `length` argument to specify how much to unmap? This would allow to partially unmap regions as fits, much like munmap() does on UNIX systems?

wipawel replied:
Uh, yes that's true. I will change this as recommended.

minipli-oss replied:
As noted throughout the series, this should be a follow-up change.

minipli-oss commented on tests/unittests.c:
```diff
@@ -98,7 +99,12 @@ static unsigned long __user_text test_user_task_func2(void *arg) {
+        ud2();
+    }
 
+    va = mmap(_ptr(0xfff80000), PAGE_ORDER_4K);
```
Can you please extend the test to add bigger and unaligned mappings as well, like doing a 2M+32k mapping at address 0xfff7fc000, which would need to be split up into 2 x 4k mappings, one 2MB mapping and yet another 2 x 4k mappings.

wipawel replied:
Ok, good idea.

minipli-oss replied:
As the current API handled single page mappings only, this should be part of a follow-up change that wires up `v[un]map_range()` for userland.

wipawel committed:
arch,pt: minor cleanup of pagetable.h header file

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
arch,pt: unify and standardize vmap API interface

Summary of the changes:
  * Get rid of kmap() API - it was confusing and non-standard
  * Add set of generic vmap APIs and make them public
      - These functions allow to specify all parameters
          * All level PT flags
          * Address space (via CR3 pointer)
  * Add set of order-specific (4K, 2M, 1G) public vmap APIs
      - Use default PT flags for non-last level
      - Allow to propagate user bit on request
  * Add separate sets of APIs for kernel and user address spaces
  * Move all interface definitions from page.h to pagetable.h

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
arch,pt: add vmap_range() and vunmap_range() API

This API allows to automatically map (or unmap) a range of physical
addresses specified with a start address and length.
Both vmap_range() and vunmap_range() allow to automatically map
the physical address range into multiple different memory areas
specified via vmap_flags parameter.

Support memory areas via vmap_flags:
  * VMAP_IDENT - kernel address space, identity mapping
  * VMAP_KERNEL - high kernel addresses (small area)
  * VMAP_KERNEL_MAP - low kernel addresses (large area)
  * VMAP_KERNEL_USER - kernel address space, but user memory area
  * VMAP_KERNEL_USER_ACCESS - kernel address space, but user memory
                              area with user access permission
  * VMAP_USER - user address space, user memory area
  * VMAP_USER_IDENT - user address space, identity mapping, no user
                      access
  * VMAP_USER_KERNEL - user address space, high kernel addresses, no
                       user access
  * VMAP_USER_KERNEL_MAP - user address space, low kernel addresses,
                           no user access

The vmap_range() tries to use as many huge pages as possible for the
mapping.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
mm,vmm: improve get_free_pages() implementation

Use vmap_range() to handle multiple mapping areas automatically
and consistently.

Standardize rules for virtual address returned by get_free_pages().

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -308,31 +308,114 @@ static void *_vmap(cr3_t *cr3_ptr, void *va, mfn_t mfn, unsigned int order,
+    unsigned long l4_flags = L4_PROT;
+
+    if (propagate_user)
+        l4_flags |= (l3_flags & _PAGE_USER);
```
No need for the brackets (as in all other such constructs following below).

wipawel replied:
Will fix. Thanks.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -308,31 +308,114 @@ static void *_vmap(cr3_t *cr3_ptr, void *va, mfn_t mfn, unsigned int order,
-    unsigned long _va = _ul(va) & PAGE_ORDER_TO_MASK(order);
+void *vmap_1g(cr3_t *cr3_ptr, void *va, mfn_t mfn, unsigned long l3_flags,
+              bool propagate_user) {
+    unsigned long _va = _ul(va) & PAGE_ORDER_TO_MASK(PAGE_ORDER_1G);
```
Should we `return -EINVAL` if `va & (PAGE_ORDER_TO_MASK(...) - 1)` or is the "automatic alignment" of `va` deliberate?

Same for `vmap_2m()`, `vmap_4k()`.

wipawel replied:
The automatic alignment of `va` is indeed deliberate for the order specific functions (`vmap_4k`, vmap_2m`, `vmap_1g`). But I have just noticed that I placed it also in the more generic function `vmap`. I will modify it to return the `-EINVAL` there. Thanks.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -537,6 +620,235 @@ frame_t *find_user_va_frame(const void *va) {
+                               flags & ~_PAGE_USER, false))
+            return err;
+    }
+
```
General remark: Should we unwind from errors instead of leaving partial mappings behind in case one of the latter ones failed?

wipawel replied:
I was thinking about it too, but decided it is too much work for very little gain as the errors are (supposed to be) very rare and typically fatal. So I decided to return an error and leave the ultimate decision to the caller what to do. 
But, I will add a comment saying this might be needed in the future.

minipli-oss reviewed (dismissed):
Great work, Pawel. Only a few nits and questions left, but really no more code changes required.

The comment regarding only being able to unmap single pages via SYSCALL_MUNMAP still applies. But I guess, now with `v[un]map_range()` available, it's easy to extend the functionality to cover multiple pages. But this can (and should!) be a follow-up change.

Thanks again for the hard work!

wipawel commented:
> Great work, Pawel. Only a few nits and questions left, but really no more code changes required.
> 
> The comment regarding only being able to unmap single pages via SYSCALL_MUNMAP still applies. But I guess, now with `v[un]map_range()` available, it's easy to extend the functionality to cover multiple pages. But this can (and should!) be a follow-up change.
> 
> Thanks again for the hard work!

Thanks very much Mathias!

I created a dedicated issue for the mmap/munmap usermode extension: #332. This should also depend on another issue: #331, but in general should be much easier to implement now.

minipli-oss reviewed (approved):
LGTM!
